src/psyclone/psyad/tl2ad.py

```diff
@@ -98,9 +155,407 @@ def generate_adjoint(tl_psyir):
+            "enclosing module.")
+
+    # First check that there's only one routine and that it's not a Program.
+    routines = tl_psyir.walk(Routine)
+
+    if len(routines) != 1:
+        raise NotImplementedError(
+            "The supplied Fortran must contain one and only one subroutine "
+            "but found: {0}".format([sub.name for sub in routines]))
+
+    tl_kernel = routines[0]
+
+    if tl_kernel.is_program:
+        raise NotImplementedError(
+            "Generation of a test harness for a kernel defined as a Program "
+            "(as opposed to a Subroutine) is not currently supported. (Found "
+            "'{0}' which is a Program.)".format(tl_kernel.name))
+
+    # First Container is a FileContainer and that's not what we want
+    container = tl_psyir.walk(Container)[1]
+
+    # Get the Container and Routine names from the PSyIR of the adjoint.
+    adjoint_kernel_name = ad_psyir.walk(Routine)[0].name
+    adjoint_module_name = ad_psyir.walk(Container)[1].name
+
+    # Create a symbol for the TL kernel
+    csym = ContainerSymbol(container.name)
+    symbol_table.add(csym)
+    tl_kernel_sym = tl_kernel.symbol_table.lookup(tl_kernel.name).copy()
+    tl_kernel_sym.interface = GlobalInterface(csym)
+    symbol_table.add(tl_kernel_sym)
+
+    # Create a symbol for the adjoint kernel
+    adj_container = ContainerSymbol(adjoint_module_name)
+    symbol_table.add(adj_container)
+    adj_kernel_sym = symbol_table.new_symbol(
+        adjoint_kernel_name, symbol_type=RoutineSymbol,
+        interface=GlobalInterface(adj_container))
+
+    # Create a symbol to hold the extent of any test arrays. This is done here
+    # to avoid any clashes with any of the container and kernel names.
+    dim_size_sym = symbol_table.new_symbol("array_extent",
+                                           symbol_type=DataSymbol,
+                                           datatype=INTEGER_TYPE,
+                                           constant_value=array_dim_size)
+
+    # Create symbols for the results of the inner products
+    inner1 = symbol_table.new_symbol("inner1", symbol_type=DataSymbol,
+                                     datatype=REAL_DOUBLE_TYPE)
+    inner2 = symbol_table.new_symbol("inner2", symbol_type=DataSymbol,
+                                     datatype=REAL_DOUBLE_TYPE)
+    # Create symbol for result of the diff of the inner products
+    diff_sym = symbol_table.new_symbol("abs_diff", symbol_type=DataSymbol,
+                                       datatype=REAL_DOUBLE_TYPE)
+
+    # Identify any arguments to the kernel that are used to dimension other
+    # arguments.
+    integer_scalars = []
+    for arg in tl_kernel.symbol_table.argument_datasymbols:
+        if arg.is_scalar and (arg.datatype.intrinsic ==
+                              ScalarType.Intrinsic.INTEGER):
+            integer_scalars.append(arg)
+    dimensioning_args = set()
+    for arg in tl_kernel.symbol_table.argument_datasymbols:
```

Comment:
I'm not suggesting this now but this feels like a nice generic bit of functionality that could be added to Routine, or perhaps a separate set of utilities.

Reply:
Good point, worth bearing in mind. Probably one for the symbol table?